politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go

```diff
@@ -0,0 +1,309 @@
+// votes cache.
+func (a *activeVotes) CommitmentAddrs(token []byte, tickets []string) map[string]commitmentAddr {
+	if len(tickets) == 0 {
+		return map[string]commitmentAddr{}
+	}
+
+	a.RLock()
+	defer a.RUnlock()
+
+	t := hex.EncodeToString(token)
+	av, ok := a.activeVotes[t]
+	if !ok {
+		return nil
+	}
+
+	ca := make(map[string]commitmentAddr, len(tickets))
+	for _, v := range tickets {
+		addr, ok := av.Addrs[v]
+		if ok {
+			ca[v] = commitmentAddr{
+				addr: addr,
+			}
+		}
+	}
+
+	return ca
+}
+
+// Tally returns the tally of the cast votes for each vote option in an active
+// vote. The returned map is a map[votebit]tally. An empty map is returned if
+// the requested token is not in the active votes cache.
+func (a *activeVotes) Tally(token string) map[string]uint32 {
+	a.RLock()
+	defer a.RUnlock()
+
+	tally := make(map[string]uint32, 16)
+	av, ok := a.activeVotes[token]
+	if !ok {
+		return tally
+	}
+	for _, votebit := range av.CastVotes {
+		tally[votebit]++
+	}
+	return tally
+}
+
+// AddCastVote adds a cast ticket vote to the active votes cache. This function
+// will panic if the provided token does not correspond to a record in the
+// active votes cache.
+func (a *activeVotes) AddCastVote(token, ticket, votebit string) {
+	a.Lock()
+	defer a.Unlock()
+
+	av, ok := a.activeVotes[token]
+	if !ok {
+		// This should not happen
+		e := fmt.Sprintf("active vote not found %v", token)
+		panic(e)
+	}
+
+	av.CastVotes[ticket] = votebit
+}
+
+// AddCommitmentAddrs adds commitment addresses to the cache for a record. This
+// function will panic if the provided token does not correspond to a record in
+// the active votes cache.
+func (a *activeVotes) AddCommitmentAddrs(token string, addrs map[string]commitmentAddr) {
+	a.Lock()
+	defer a.Unlock()
+
+	av, ok := a.activeVotes[token]
+	if !ok {
+		// This should not happen
+		e := fmt.Sprintf("active vote not found %v", token)
+		panic(e)
+	}
+
+	for ticket, v := range addrs {
+		if v.err != nil {
+			log.Errorf("Commitment address error %v %v %v",
+				token, ticket, v.err)
+			continue
+		}
+		av.Addrs[ticket] = v.addr
+	}
+}
+
+// Del deletes an active vote from the active votes cache.
+func (a *activeVotes) Del(token string) {
+	a.Lock()
+	defer a.Unlock()
+
+	delete(a.activeVotes, token)
+
+	log.Debugf("Active votes del %v", token)
+}
+
+// Add adds a active vote to the active votes cache.
+//
+// This function should NOT be used directly. The activeVotesAdd function,
+// which also kicks of an async job to fetch the commitment addresses for this
+// active votes entry, should be used instead.
+func (a *activeVotes) Add(vd ticketvote.VoteDetails) {
+	a.Lock()
+	defer a.Unlock()
+
+	token := vd.Params.Token
+	a.activeVotes[token] = activeVote{
+		Details:   &vd,
+		CastVotes: make(map[string]string, 40960), // Ticket pool size
+		Addrs:     make(map[string]string, 40960), // Ticket pool size
+	}
+
+	log.Debugf("Active votes add %v", token)
```

Comment:
Don't do this inside mutex. The makes are questionable but I can live with those. Don't see the point in this case to create a token variable either.